File: lib/php/visualplatform-authenticate.php
#!/usr/bin/env php
<?

# Instructions:
#
# Register your application by following the instruction on 
#   http://community.23video.com/api/
#
# Call the script with your key and secret:
#  ./visualplatform-authenticate.php <consumer_key> <consumer_secret>
#
# Upon successful authentication, you'll get a visualplatform.config.php
# file with your OAuth credentials


// Get key/token from args
$args = $_SERVER['argv'];
$consumer_key = $args[1];
$consumer_secret = $args[2];
$domain = "api.visualplatform.net";


// Load dependencies
include_once 'HTTP/OAuth/Consumer.php';
require_once 'HTTP/OAuth.php';
require_once 'HTTP/OAuth/Consumer/Request.php';
require_once 'HTTP/Request2.php';

// Set up HTTP request
$httpRequest = new HTTP_Request2;
$httpRequest->setHeader('Accept-Encoding', '.*');
$request = new HTTP_OAuth_Consumer_Request;
$request->accept($httpRequest);

// Set up OAuth consumer
$consumer = new HTTP_OAuth_Consumer($consumer_key, $consumer_secret);
$consumer->accept($request);

// 1. Get request token
$consumer->getRequestToken("http://".$domain."/oauth/request_token", "oob", array(), "GET");

// 2. Build authorize url and redirect the user
$authorize_url = $consumer->getAuthorizeUrl("http://".$domain."/oauth/authorize");
echo("Please visit the following URL in your browser to authorize your application, then enter the 4 character security code when done:\n\n  " . $authorize_url . "\n\nVerification code: ");

// 3. Get an OAuth verifier (this step is only necessary when a callback url hasn't been specified)
$handle = @fopen("php://stdin", "r");
$oauth_verifier = trim(fgets($handle));
fclose($handle);

// 4. Exchange the request token for an access token
//    (He're we're reading extra data from the access token request to 
//    determine the domain and user_id of the authenticated user. 
//    Unfortunately this PHP OAuth library doesn't return the excess
//    data from the access token request, so we'll need to handle the
//    request manually.)

$response = $consumer->sendRequest("http://".$domain."/oauth/access_token", array("oauth_verifier" => $oauth_verifier), "GET");
$data     = $response->getDataFromBody();

if (empty($data['oauth_token']) || empty($data['oauth_token_secret'])) {
    throw new HTTP_OAuth_Consumer_Exception_InvalidResponse(
       'Failed getting token and token secret from response', $response
    );
}

$consumer->setToken($data['oauth_token']);
$consumer->setTokenSecret($data['oauth_token_secret']);


// 5. Save credentials to config file
$handle = @fopen("visualplatform.config.php", "w");
$_conf = array(
                'domain'       => $data['domain'],
                'user_id'      => $data['user_id'],
                'key'          => $consumer->getKey(),
                'secret'       => $consumer->getSecret(),
                'token'        => $consumer->getToken(),
                'token_secret' => $consumer->getTokenSecret()
        );
fwrite($handle, '<? $visualplatform_config = unserialize(\'' . serialize($_conf) . '\'); ?>');
fclose($handle);



?>

File: lib/ruby/visualplatform-authenticate.rb
#!/usr/bin/env ruby

# Instructions:
#
# Register your application by following the instruction on
#   http://community.23video.com/api/
#
# Upon successful registration, you'll get a yaml file with oauth credentials
# Pass these values on the command line as --key (-k) and --secret (-s) then
# follow the instructions.
#
# This script is a cheap copy of the oauth script in yammer4r.

require 'optparse'
require 'rubygems'
require 'oauth'

options = {
  :outfile  => 'oauth.yml'
}

VP_OAUTH = "http://api.visualplatform.net"

ARGV.options do |o|
  script_name = File.basename($0)

  o.set_summary_indent('  ')
  o.banner =    "Usage: #{script_name} [OPTIONS]"
  o.define_head "Create a yaml file for Visual Platform OAuth"
  o.separator   ""
  o.separator   "[-k] and [-s] options are mandatory"
  o.separator   ""

  o.on("-o", "--outfile=[val]", String,
       "Yaml output file",
       "Default: #{options[:outfile]}")     { |opts| options[:outfile] = opts }
  o.on("-k", "--key=val", String,
       "Consumer key for Visual Platform app")       { |key| options[:key] = key}
  o.on("-s", "--secret=val", String,
       "Consumer secret for Visual Platform app")    { |secret| options[:secret] = secret}

  o.separator ""

  o.on_tail("-h", "--help", "Show this help message.") { puts o; exit }
  o.parse!
end

unless options[:key] && options[:secret]
  raise ArgumentError, "Must supply consumer key and secret (use -h for help)"
end

consumer      = OAuth::Consumer.new options[:key], options[:secret], {:site => VP_OAUTH}
request_token = consumer.get_request_token

puts "Please visit the following URL in your browser to authorize your application, then enter the 4 character security code when done: #{request_token.authorize_url}"
oauth_verifier =  gets
response = consumer.token_request(consumer.http_method,
                                  (consumer.access_token_url? ? consumer.access_token_url : consumer.access_token_path),
                                  request_token,
                                  {},
                                  :oauth_verifier =>  oauth_verifier.chomp)
access_token = OAuth::AccessToken.new(consumer,response[:oauth_token],response[:oauth_token_secret])

oauth_yml = <<-EOT
consumer:
  key: #{options[:key]}
  secret: #{options[:secret]}
access:
  token: #{access_token.token}
  secret: #{access_token.secret}
api:
  debug: true
  endpoint: http://#{response['domain']}
  domain: #{response['domain']}
  user_id: #{response['user_id']}
  oauth: #{VP_OAUTH}
EOT

File.open(options[:outfile], "w") do |f|
  f.write oauth_yml
end
